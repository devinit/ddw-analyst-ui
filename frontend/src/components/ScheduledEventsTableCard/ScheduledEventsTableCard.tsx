import React, { FunctionComponent, ReactNode, useEffect, useState } from 'react';
import { Card } from 'react-bootstrap';
import { Dimmer, Loader } from 'semantic-ui-react';
import { fetchData, getScheduledEventsByPage, LIMIT } from '.';
import { ScheduledEvent } from '../../types/scheduledEvents';
import { PaginationRow } from '../PaginationRow';
import { ScheduledEventsTable } from '../ScheduledEventsTable';

export const ScheduledEventsTableCard: FunctionComponent = () => {
  const [scheduledEvents, setScheduledEvents] = useState<ScheduledEvent[]>([]);
  const [loading, setLoading] = useState(true);
  const [count, setCount] = useState(0);
  const [pageCount, setPageCount] = useState(0);
  const [currentPage, setCurrentPage] = useState(1);
  const [selectedPage, setSelectedPage] = useState(0);

  useEffect(() => {
    fetchData().then((result) => {
      setScheduledEvents(result.data);
      setLoading(false);
      setCount(result.data.length);
      setPageCount(Math.ceil(result.data.length / LIMIT));
    });
  }, []);

  const handlePageChange = (page: { selected: number }): void => {
    setSelectedPage(page.selected);
    if (page.selected === selectedPage + 1) {
      setCurrentPage((currentPage) => currentPage + 1);
    } else if (page.selected === selectedPage - 1) {
      setCurrentPage((currentPage) => currentPage - 1);
    } else {
      setCurrentPage(() => page.selected + 1);
    }
  };

  const renderPagination = (): ReactNode => {
    return count === 0 ? (
      'No Data'
    ) : (
      <PaginationRow
        limit={LIMIT}
        count={count}
        pageCount={pageCount}
        onPageChange={handlePageChange}
      />
    );
  };

  return (
    <React.Fragment>
      <Dimmer active={loading} inverted>
        <Loader content="Loading" />
      </Dimmer>
      <Card className="col-md-12">
        <Card.Header className="card-header-rose card-header-icon">
          <Card.Header className="card-icon">
            <i className="material-icons">schedule</i>
          </Card.Header>
          <h4 className="card-title">Scheduled Events</h4>
        </Card.Header>
        <Card.Body>
          <ScheduledEventsTable
            currentPage={currentPage}
            pageLimit={LIMIT}
            events={getScheduledEventsByPage(currentPage, scheduledEvents)}
          />
          {renderPagination()}
        </Card.Body>
      </Card>
    </React.Fragment>
  );
};
