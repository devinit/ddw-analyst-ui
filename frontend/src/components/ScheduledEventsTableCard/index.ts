export * from './ScheduledEventsTableCard';
export * from './utils';
