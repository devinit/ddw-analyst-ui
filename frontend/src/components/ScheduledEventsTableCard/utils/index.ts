import axios from 'axios';
import * as localForage from 'localforage';
import { ScheduledEvent } from '../../../types/scheduledEvents';
import { api, localForageKeys } from '../../../utils';

const BASEPATH = api.routes.VIEW_SCHEDULED_EVENTS;

export const LIMIT = 5;

export const fetchData = async (): Promise<any> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `token ${token}`,
  };

  return await axios(`${BASEPATH}`, { headers });
};

export const getScheduledEventsByPage = (
  currentPage: number,
  data: ScheduledEvent[],
): ScheduledEvent[] => {
  const begin = (currentPage - 1) * 5;
  const end = begin + 5;

  return data.slice(begin, end);
};
