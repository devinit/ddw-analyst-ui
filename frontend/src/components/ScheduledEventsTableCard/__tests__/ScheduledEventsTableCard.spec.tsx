/**
 * @jest-environment jsdom
 */
import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { ScheduledEventsTableCard } from '../ScheduledEventsTableCard';

describe('<ScheduledEventsTableCard', () => {
  afterEach(cleanup);

  it('should match snapshot', () => {
    const { container } = render(<ScheduledEventsTableCard />);
    expect(container).toMatchSnapshot();
  });
});
