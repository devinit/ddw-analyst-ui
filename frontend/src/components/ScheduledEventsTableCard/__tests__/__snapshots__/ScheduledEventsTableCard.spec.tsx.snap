// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScheduledEventsTableCard should match snapshot 1`] = `
<div>
  <div
    class="ui active transition visible inverted dimmer"
    style="display: flex;"
  >
    <div
      class="content"
    >
      <div
        class="ui text loader"
      >
        Loading
      </div>
    </div>
  </div>
  <div
    class="col-md-12 card"
  >
    <div
      class="card-header-danger card-header-icon card-header"
    >
      <div
        class="card-icon card-header"
      >
        <i
          class="material-icons"
        >
          schedule
        </i>
      </div>
      <h4
        class="card-title"
      >
        Scheduled Events
      </h4>
    </div>
    <div
      class="card-body"
    >
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped"
        >
          <thead>
            <tr>
              <th
                class="text-center"
              >
                #
              </th>
              <th>
                Name
              </th>
              <th>
                Description
              </th>
              <th>
                Enabled
              </th>
              <th>
                Start Date
              </th>
              <th>
                Interval
              </th>
              <th>
                Actions
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      No Data
    </div>
  </div>
</div>
`;
